### ISSUE TITLE
Allow implicit conversion between integer types

### ISSUE THREAD

**Issue by srknzl** [hazelcast/hazelcast#19042]
Recently, implicit conversion between integer types is removed. This can cause problems in languages that do not have all the integer types. For example, in Node.js there is only one number type which is like a double. We use a 3rd party library called long that maps to BIGINT type. Native numbers are sent as double by default, and it is configurable via the `defaultNumberType` configuration. 

If long objects can be implicitly converted to all other integer types, users won't have to write explicit casts in SQL queries.

The following query is taken from a client test, it was working recently, but not working now: (the map includes java.lang.Byte values)

```SQL
`SELECT * FROM ${mapName} WHERE this > ? AND this < ? ORDER BY __key ASC`,
params: [long.fromNumber(10), long.fromNumber(16)]
```

instead it became:

```SQL
`SELECT * FROM ${mapName} WHERE this > CAST(? AS TINYINT) AND this < CAST(? AS TINYINT) ORDER BY __key ASC`, 
params:[10, 16](no need to use long anymore, we are casting doubles here)
```

cc @viliam-durina

**Comment by Fly-Style** [comment]
Hi Serkan @srknzl, thanks for the report.
Honestly, it's our deliberate decision to have strict SQL dialect without any hidden implicit casts. We are mostly following PostgreSQL-like approach for it.

**Comment by srknzl** [comment]
Ok, if you can revert the change it will be better for some non-java clients; if you won't, you can close the issue

**Comment by viliam-durina** [comment]
@Fly-Style We do allow some implicit casts, e.g. from TINYINT to INTEGER. I think it makes sense to allow also the reverse conversion, however it must check for out-of-range values. However, we won't allow conversion from INT to DOUBLE, even though you can assign a double literal into an INT column, but it would create weird behavior when the parameter is used in an expression. For example:
```sql
SELECT * FROM foo WHERE intColumn=?
```
If the param value is a `double` value, it would first get converted to `int` (and lose fractional part) and then the `int` value will be used in the expression. However if a literal was provided, then `intColumn` would be cast to DOUBLE and then compared as doubles.

I implemented a fix in #20221.
